Optimize conversions from NSArray / CFArrayRef to Vector

cdumez opened the pull request:
#### d8e283330a3152118a1a0832dbd23b5a542b16f0
<pre>
Optimize conversions from NSArray / CFArrayRef to Vector
<a href="https://bugs.webkit.org/show_bug.cgi?id=291769">https://bugs.webkit.org/show_bug.cgi?id=291769</a>

Reviewed by Darin Adler.

Optimize conversions from NSArray / CFArrayRef to Vector by leveraging the
`Vector(size_t, const Functor&amp; valueGenerator)` constructor. It avoids
capacity checks when appending elements.

* Source/WTF/wtf/Vector.h:
(WTF::Vector::Vector):
* Source/WTF/wtf/cf/VectorCF.h:
(WTF::makeVector):
* Source/WTF/wtf/cocoa/VectorCocoa.h:
(WTF::makeVector):

Canonical link: <a href="https://commits.webkit.org/294292@main">https://commits.webkit.org/294292@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/addfe039bf2a9f980370f5f7d415c47718ed76ef

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101355 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21019 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11324 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106508 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51984 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21327 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29513 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/77208 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34234 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104362 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16455 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91549 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/57549 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16275 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9559 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51332 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/94024 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86156 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9635 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108860 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/99965 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28484 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/20957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/86179 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28846 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87751 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85736 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21812 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30456 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8168 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22600 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28415 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33694 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/123590 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28226 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34376 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31546 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29785 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -745,8 +745,9 @@ class Vector : private VectorBuffer<T, inlineCapacity, Malloc> {
                     unsafeAppendWithoutCapacityCheck(WTFMove(*item));
                 else
-                    return;
+                    continue;
```
Are you sure no one is accidentally depending on "first `nullopt` stops the iteration" behavior?

cdumez replied:
I will try and find out. It would result in correct behavior but I guess your point is that it may be slower.

cdumez replied:
Oh actually, this can lead to trouble if the lambda doesn't expect to get called with future indexes. I will rework the PR.

darinadler reviewed (approved):
I’m guessing not, but is there a measurable speedup?

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -55,6 +55,8 @@ namespace WTF {
 DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(Vector);
 DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(VectorBuffer);
 
+enum class NulloptBehavior : bool { Ignore, Abort };
```
Annoying that we have to add this but I think we do! Or maybe we could use a `std::variant` with a type other than `std::nullopt_t` for the abort behavior.

Should we make ignore the default?

cdumez replied:
I liked the idea of returning a std::optional since this is what we do for `compactMap()` for example. I feel that using `std::variant` would add some complexity. However, I think it would make sense to make "Ignore" the default.

cdumez committed:
Optimize conversions from NSArray / CFArrayRef to Vector
https://bugs.webkit.org/show_bug.cgi?id=291769

Reviewed by Darin Adler.

Optimize conversions from NSArray / CFArrayRef to Vector by leveraging the
`Vector(size_t, const Functor& valueGenerator)` constructor. It avoids
capacity checks when appending elements.

* Source/WTF/wtf/Vector.h:
(WTF::Vector::Vector):
* Source/WTF/wtf/cf/VectorCF.h:
(WTF::makeVector):
* Source/WTF/wtf/cocoa/VectorCocoa.h:
(WTF::makeVector):

Canonical link: https://commits.webkit.org/294292@main
